tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -733,351 +735,389 @@ getParamParentNameOffset(const ValueDecl *VD, SourceLoc Cursor) {
+            Target, PrintOptions::printModuleInterface(
+                        Invoc.getFrontendOptions().PrintFullConvention));
+        InSynthesizedExtension = Analyzer.isInSynthesizedExtension(VD);
+      }
     }
   }
-  unsigned USREnd = SS.size();
 
-  unsigned TypenameBegin = SS.size();
-  llvm::raw_svector_ostream OS(SS);
-  PrintOptions Options;
-  Options.PrintTypeAliasUnderlyingType = true;
-  VD->getInterfaceType().print(OS, Options);
+  bool isValid() { return VD != nullptr; }
+};
 
-  unsigned TypenameEnd = SS.size();
+/// Delays constructing `Assign` from `Buffer` until destructed, ie. assumes
+/// that the buffer will no longer be modified after the destructor is called.
+///
+/// `Printer` returns whether or not `Assign` should be assigned, in case it
+/// needs to be overridden.
+class DelayedAssign {
+  StringRef &Assign;
+  SmallVectorImpl<char> &Buffer;
+  unsigned Begin;
+  unsigned End;
+  bool ShouldAssign;
 
-  unsigned MangledTypeStart = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printDeclTypeUSR(VD, OS);
+public:
+  DelayedAssign(StringRef &Assign, SmallVectorImpl<char> &Buffer,
+                llvm::function_ref<bool(llvm::raw_svector_ostream &)> Printer)
+      : Assign(Assign), Buffer(Buffer), Begin(Buffer.size()) {
+    llvm::raw_svector_ostream OS(Buffer);
+    ShouldAssign = Printer(OS);
+    End = Buffer.size();
   }
-  unsigned MangledTypeEnd = SS.size();
 
-  unsigned MangledContainerTypeStart = SS.size();
-  if (ContainerTy && !ContainerTy->hasArchetype()) {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printTypeUSR(ContainerTy, OS);
+  StringRef unstableRef() {
+    if (!ShouldAssign)
+      return Assign;
+    return StringRef(Buffer.begin() + Begin, End - Begin);
   }
-  unsigned MangledContainerTypeEnd = SS.size();
-
-  // If VD is the syntehsized property wrapper backing storage (_foo) or
-  // projected value ($foo) of a property (foo), use that property's
-  // documentation instead.
 
-  unsigned DocCommentBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    ide::getDocumentationCommentAsXML(OriginalProperty, OS);
+  ~DelayedAssign() {
+    if (ShouldAssign)
+      Assign = unstableRef();
   }
-  unsigned DocCommentEnd = SS.size();
+};
 
-  unsigned DeclBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    printAnnotatedDeclaration(VD, BaseType, OS);
-  }
-  unsigned DeclEnd = SS.size();
+static llvm::Error
+fillSymbolInfo(CursorSymbolInfo &Symbol, DeclInfo &DInfo,
+               ModuleDecl *MainModule, SourceLoc CursorLoc, bool AddSymbolGraph,
+               SwiftLangSupport &Lang, const CompilerInvocation &Invoc,
+               ArrayRef<ImmutableTextSnapshotRef> PreviousSnaps,
+               SmallVectorImpl<char> &StrBuffer) {
+  Symbol.DeclarationLang = SwiftLangSupport::getUIDForDeclLanguage(DInfo.VD);
+  Symbol.Kind = SwiftLangSupport::getUIDForDecl(DInfo.VD, DInfo.IsRef);
+
+  auto AssignName = DelayedAssign(
```

Comment:
Another option here (rather than DelayedAssign and DelayedStringRetriever) would be to have a single buffer + BumpPtrAllocator, with the single buffer being re-used and then everything copied into the BumpPtrAllocator. We could do that for the SmallVectors as well, which would let them be an ArrayRef again. Bit more copying... but that should be pretty cheap?

Reply:
I really wasn't happy with how this was, so I ended up doing what I just suggested. IMO what's going on is much clearer now and the copies should be quite cheap. It's a separate commit, so can pull out if you prefer the Delayed* method instead.